File: 1116-maximum-level-sum-of-a-binary-tree/1116-maximum-level-sum-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        q = deque()
        q.append(root)

        cur_level = 1
        answer = 0
        max_sum = float('-inf')      
        while q:
            q_size = len(q)
            sum = 0
            for i in range(len(q)):
                node = q.popleft()
                sum += node.val

                if node.left:
                    q.append(node.left)
                
                if node.right:
                    q.append(node.right)

            if sum > max_sum:
                max_sum = sum
                answer = cur_level
            
            cur_level += 1
        return answer
        
        

                    
                


        
        
        

File: 1116-maximum-level-sum-of-a-binary-tree/1116-maximum-level-sum-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int maxLevelSum(TreeNode root) {
        return bfs(root);
    }

    int bfs(TreeNode root) {
        Queue<TreeNode> q = new ArrayDeque<>();
        q.offer(root);
        int maxValue = 0;
        Map<Integer, Integer> map = new HashMap<>();

        int idx = 0;
        while (!q.isEmpty()) {
            int qSize = q.size();
            int result = 0;            
            for (int i=0; i<qSize; ++i) {
                TreeNode now = q.poll();

                if (now == null) {
                    continue;
                }

                result += now.val;

                if (now.left != null) {
                    q.offer(now.left);
                }

                if (now.right != null) {
                    q.offer(now.right);
                }
            }
            idx++;
            if (map.get(result) == null) {
                map.put(result, idx);
            }

            maxValue = Math.max(maxValue, result);
        }

        return map.getOrDefault(maxValue, idx);
    }
}